### ISSUE TITLE
Constant accuracy metric for LeNet-5 example

### ISSUE THREAD

**Issue by umangjpatel** [google/jax#2063]
I've been exploring the Stax layers API for experimentation. I was attempting to try the LeNet-5 architecture on examples/mnist_classifier.py script. I found that the accuracy doesn't increase after a while. What's the problem over here ?

### Code 
```python3
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import time
import itertools

import numpy.random as npr

import jax.numpy as np
from jax import jit, grad, random
from jax.experimental import optimizers
from jax.experimental import stax
from jax.experimental.stax import Dense, Relu, LogSoftmax, Conv, MaxPool, Flatten
from examples import dataset


def loss(params, batch):
    inputs, targets = batch
    preds = predict(params, inputs)
    return -np.mean(np.sum(preds * targets, axis=1))


def accuracy(params, batch):
    inputs, targets = batch
    target_class = np.argmax(targets, axis=1)
    predicted_class = np.argmax(predict(params, inputs), axis=1)
    return np.mean(predicted_class == target_class)


init_random_params, predict = stax.serial(
    Conv(out_chan=6, filter_shape=(5, 5), strides=(1, 1), padding="VALID"), Relu,
    MaxPool(window_shape=(2, 2), padding="VALID", spec="NCHW"),
    Conv(out_chan=16, filter_shape=(5, 5), strides=(1, 1), padding="VALID"), Relu,
    MaxPool(window_shape=(2, 2), padding="VALID", spec="NCHW"),
    Flatten,
    Dense(120), Relu,
    Dense(84), Relu,
    Dense(10), LogSoftmax)

if __name__ == "__main__":
    rng = random.PRNGKey(0)

    step_size = 0.001
    num_epochs = 10
    batch_size = 128
    momentum_mass = 0.9

    train_images, train_labels, test_images, test_labels = dataset.mnist()
    num_train = train_images.shape[0]
    num_complete_batches, leftover = divmod(num_train, batch_size)
    num_batches = num_complete_batches + bool(leftover)

    train_images = np.asarray(train_images.reshape(-1, 1, 28, 28))
    test_images = np.asarray(test_images.reshape(-1, 1, 28, 28))

    print(train_images.shape, train_labels.shape)
    print(test_images.shape, test_labels.shape)


    def data_stream():
        rng = npr.RandomState(0)
        while True:
            perm = rng.permutation(num_train)
            for i in range(num_batches):
                batch_idx = perm[i * batch_size:(i + 1) * batch_size]
                yield train_images[batch_idx], train_labels[batch_idx]


    batches = data_stream()

    opt_init, opt_update, get_params = optimizers.sgd(step_size)


    @jit
    def update(i, opt_state, batch):
        params = get_params(opt_state)
        return opt_update(i, grad(loss)(params, batch), opt_state)


    _, init_params = init_random_params(rng, (-1, 1, 28, 28))
    opt_state = opt_init(init_params)
    itercount = itertools.count()

    print("\nStarting training...")
    for epoch in range(num_epochs):
        start_time = time.time()
        for _ in range(num_batches):
            opt_state = update(next(itercount), opt_state, next(batches))
        epoch_time = time.time() - start_time

        params = get_params(opt_state)
        train_acc = accuracy(params, (train_images, train_labels))
        test_acc = accuracy(params, (test_images, test_labels))
        print("Epoch {} in {:0.2f} sec".format(epoch, epoch_time))
        print("Training set accuracy {}".format(train_acc))
        print("Test set accuracy {}".format(test_acc))
```

### Output
```terminal
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/jax/lib/xla_bridge.py:119: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
(60000, 1, 28, 28) (60000, 10)
(10000, 1, 28, 28) (10000, 10)
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/jax/nn/initializers.py:46: RuntimeWarning: invalid value encountered in long_scalars
  receptive_field_size = onp.prod(shape) / shape[in_axis] / shape[out_axis]

Starting training...
Epoch 0 in 1.81 sec
Training set accuracy 0.10218333452939987
Test set accuracy 0.10100000351667404
Epoch 1 in 0.67 sec
Training set accuracy 0.11236666887998581
Test set accuracy 0.11349999904632568
Epoch 2 in 0.69 sec
Training set accuracy 0.11236666887998581
Test set accuracy 0.11349999904632568
Epoch 3 in 0.63 sec
Training set accuracy 0.11236666887998581
Test set accuracy 0.11349999904632568
Epoch 4 in 0.67 sec
Training set accuracy 0.11236666887998581
Test set accuracy 0.11349999904632568
Epoch 5 in 0.62 sec
Training set accuracy 0.11236666887998581
Test set accuracy 0.11349999904632568
Epoch 6 in 0.62 sec
Training set accuracy 0.11236666887998581
Test set accuracy 0.11349999904632568
Epoch 7 in 0.62 sec
Training set accuracy 0.11236666887998581
Test set accuracy 0.11349999904632568
Epoch 8 in 0.62 sec
Training set accuracy 0.11236666887998581
Test set accuracy 0.11349999904632568
Epoch 9 in 0.66 sec
Training set accuracy 0.11236666887998581
Test set accuracy 0.11349999904632568

Process finished with exit code 0
```

**Comment by umangjpatel** [comment]
Issue has been fixed.

### Code changes:-
```python3
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import time
import itertools

import numpy.random as npr

import jax.numpy as np
from jax import jit, grad, random
from jax.experimental import optimizers
from jax.experimental import stax
from jax.experimental.stax import Dense, Relu, LogSoftmax, Conv, MaxPool, Flatten
from examples import dataset


def loss(params, batch):
    inputs, targets = batch
    preds = predict(params, inputs)
    return -np.mean(np.sum(preds * targets, axis=1))


def accuracy(params, batch):
    inputs, targets = batch
    target_class = np.argmax(targets, axis=1)
    predicted_class = np.argmax(predict(params, inputs), axis=1)
    return np.mean(predicted_class == target_class)


init_random_params, predict = stax.serial(
    Conv(out_chan=6, filter_shape=(5, 5), strides=(1, 1), padding="VALID"), Relu,
    # MaxPool(window_shape=(2, 2), padding="VALID", spec="NCHW"),
    MaxPool(window_shape=(2, 2), padding="VALID", spec="NHWC"),
    Conv(out_chan=16, filter_shape=(5, 5), strides=(1, 1), padding="VALID"), Relu,
    # MaxPool(window_shape=(2, 2), padding="VALID", spec="NCHW"),
    MaxPool(window_shape=(2, 2), padding="VALID", spec="NHWC"),
    Flatten,
    Dense(120), Relu,
    Dense(84), Relu,
    Dense(10), LogSoftmax)

if __name__ == "__main__":
    rng = random.PRNGKey(0)

    step_size = 0.001
    num_epochs = 10
    batch_size = 128
    momentum_mass = 0.9

    train_images, train_labels, test_images, test_labels = dataset.mnist()
    num_train = train_images.shape[0]
    num_complete_batches, leftover = divmod(num_train, batch_size)
    num_batches = num_complete_batches + bool(leftover)

    # train_images = np.asarray(train_images.reshape(-1, 1, 28, 28))
    train_images = np.asarray(train_images.reshape(-1, 28, 28, 1))
    # test_images = np.asarray(test_images.reshape(-1, 1, 28, 28))
    test_images = np.asarray(test_images.reshape(-1, 28, 28, 1))

    print(train_images.shape, train_labels.shape)
    print(test_images.shape, test_labels.shape)


    def data_stream():
        rng = npr.RandomState(0)
        while True:
            perm = rng.permutation(num_train)
            for i in range(num_batches):
                batch_idx = perm[i * batch_size:(i + 1) * batch_size]
                yield train_images[batch_idx], train_labels[batch_idx]


    batches = data_stream()

    opt_init, opt_update, get_params = optimizers.momentum(step_size, mass=momentum_mass)


    @jit
    def update(i, opt_state, batch):
        params = get_params(opt_state)
        return opt_update(i, grad(loss)(params, batch), opt_state)


    # _, init_params = init_random_params(rng, (-1, 1, 28, 28))
    _, init_params = init_random_params(rng, (-1, 28, 28, 1))
    opt_state = opt_init(init_params)
    itercount = itertools.count()

    print("\nStarting training...")
    for epoch in range(num_epochs):
        start_time = time.time()
        for _ in range(num_batches):
            opt_state = update(next(itercount), opt_state, next(batches))
        epoch_time = time.time() - start_time

        params = get_params(opt_state)
        train_acc = accuracy(params, (train_images, train_labels))
        test_acc = accuracy(params, (test_images, test_labels))
        print("Epoch {} in {:0.2f} sec".format(epoch, epoch_time))
        print("Training set accuracy {}".format(train_acc))
        print("Test set accuracy {}".format(test_acc))
```

### Output
```terminal
/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/jax/lib/xla_bridge.py:119: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
(60000, 28, 28, 1) (60000, 10)
(10000, 28, 28, 1) (10000, 10)

Starting training...
Epoch 0 in 17.72 sec
Training set accuracy 0.9172333478927612
Test set accuracy 0.9259999990463257
Epoch 1 in 12.61 sec
Training set accuracy 0.9487000107765198
Test set accuracy 0.9509000182151794
Epoch 2 in 13.22 sec
Training set accuracy 0.9577666521072388
Test set accuracy 0.9595999717712402
Epoch 3 in 13.21 sec
Training set accuracy 0.9717666506767273
Test set accuracy 0.9724000096321106
Epoch 4 in 13.01 sec
Training set accuracy 0.9754499793052673
Test set accuracy 0.9740999937057495
Epoch 5 in 13.30 sec
Training set accuracy 0.9771166443824768
Test set accuracy 0.9753000140190125
Epoch 6 in 13.28 sec
Training set accuracy 0.9786666631698608
Test set accuracy 0.9763000011444092
Epoch 7 in 13.20 sec
Training set accuracy 0.9834499955177307
Test set accuracy 0.9801999926567078
Epoch 8 in 13.43 sec
Training set accuracy 0.9843000173568726
Test set accuracy 0.9824000000953674
Epoch 9 in 13.04 sec
Training set accuracy 0.9833499789237976
Test set accuracy 0.9815999865531921

Process finished with exit code 0
```
